File: examples/weather/weather.py
#!/usr/bin/env python3
"""weather — an ACLI-compliant weather CLI for agent consumption.

A complete example showing how to build an agent-friendly CLI tool
using the ACLI Python SDK. Demonstrates:
- Auto-injected --output (no need to declare it on every command)
- Auto-injected --dry-run on non-idempotent commands
- NDJSON streaming for long-running operations
- Actionable error messages with hints
- Semantic exit codes
"""

from __future__ import annotations

import random
import time
from typing import Any

import typer

from acli import (
    ACLIApp,
    ExitCode,
    InvalidArgsError,
    NotFoundError,
    OutputFormat,
    acli_command,
    emit,
    emit_progress,
    emit_result,
    success_envelope,
)

app = ACLIApp(name="weather", version="1.0.0")

# ── Simulated data ────────────────────────────────────────────────────────────

CITIES: dict[str, dict[str, Any]] = {
    "london": {"lat": 51.5, "lon": -0.1, "country": "GB"},
    "paris": {"lat": 48.9, "lon": 2.3, "country": "FR"},
    "tokyo": {"lat": 35.7, "lon": 139.7, "country": "JP"},
    "new-york": {"lat": 40.7, "lon": -74.0, "country": "US"},
    "sydney": {"lat": -33.9, "lon": 151.2, "country": "AU"},
}

ALERTS: list[dict[str, Any]] = [
    {"city": "tokyo", "type": "typhoon_warning", "severity": "high", "message": "Typhoon approaching"},
    {"city": "london", "type": "fog_advisory", "severity": "low", "message": "Dense fog expected"},
]

FAVORITES: list[str] = []


def _get_weather(city: str) -> dict[str, Any]:
    """Generate simulated weather data for a city."""
    random.seed(hash(city))
    return {
        "city": city,
        "country": CITIES[city]["country"],
        "temperature_c": round(random.uniform(-5, 35), 1),
        "humidity_pct": random.randint(30, 95),
        "wind_kph": round(random.uniform(0, 50), 1),
        "condition": random.choice(["sunny", "cloudy", "rainy", "snowy", "windy"]),
        "coordinates": {"lat": CITIES[city]["lat"], "lon": CITIES[city]["lon"]},
    }


# ── Commands ──────────────────────────────────────────────────────────────────
# Note: --output is auto-injected by @acli_command — no need to declare it.
# Note: --dry-run is auto-injected on idempotent=False commands.


@app.command()
@acli_command(
    examples=[
        ("Get weather for London", "weather get --city london"),
        ("Get weather for Tokyo in JSON", "weather get --city tokyo --output json"),
    ],
    idempotent=True,
    see_also=["forecast", "alerts"],
)
def get(
    city: str = typer.Option(..., help="City name (lowercase, hyphenated). type:string"),
    units: str = typer.Option("metric", help="Unit system. type:enum[metric|imperial]"),
    output: OutputFormat = typer.Option(
        OutputFormat.text, help="Output format. type:enum[text|json|table]"
    ),
) -> None:
    """Get current weather for a city.

    Returns temperature, humidity, wind speed, and conditions for the
    specified city. Use --units imperial for Fahrenheit/mph.
    """
    start = time.time()
    city = city.lower()
    if city not in CITIES:
        raise NotFoundError(
            f"Unknown city: '{city}'",
            hint=f"Available cities: {', '.join(sorted(CITIES))}",
            docs=".cli/examples/get.sh",
        )

    data = _get_weather(city)
    if units == "imperial":
        data["temperature_f"] = round(data["temperature_c"] * 9 / 5 + 32, 1)
        data["wind_mph"] = round(data["wind_kph"] * 0.621, 1)

    emit(success_envelope("get", data, version="1.0.0", start_time=start), output)


@app.command()
@acli_command(
    examples=[
        ("Get 3-day forecast for Paris", "weather forecast --city paris --days 3"),
        ("Get 7-day forecast in JSON", "weather forecast --city london --days 7 --output json"),
    ],
    idempotent=True,
    see_also=["get", "alerts"],
)
def forecast(
    city: str = typer.Option(..., help="City name (lowercase, hyphenated). type:string"),
    days: int = typer.Option(3, help="Number of forecast days (1-7). type:int"),
    output: OutputFormat = typer.Option(
        OutputFormat.text, help="Output format. type:enum[text|json|table]"
    ),
) -> None:
    """Get multi-day weather forecast for a city.

    Returns daily high/low temperatures and conditions for the next N days.
    """
    start = time.time()
    city = city.lower()
    if city not in CITIES:
        raise NotFoundError(
            f"Unknown city: '{city}'",
            hint=f"Available cities: {', '.join(sorted(CITIES))}",
        )
    if not 1 <= days <= 7:
        raise InvalidArgsError(
            f"Days must be between 1 and 7, got {days}",
            hint="Use --days with a value from 1 to 7",
        )

    random.seed(hash(city))
    daily = []
    for day in range(days):
        daily.append({
            "day": day + 1,
            "high_c": round(random.uniform(5, 35), 1),
            "low_c": round(random.uniform(-5, 20), 1),
            "condition": random.choice(["sunny", "cloudy", "rainy", "snowy"]),
            "precipitation_pct": random.randint(0, 100),
        })

    data = {"city": city, "country": CITIES[city]["country"], "days": daily}
    emit(success_envelope("forecast", data, version="1.0.0", start_time=start), output)


@app.command()
@acli_command(
    examples=[
        ("Check all active alerts", "weather alerts"),
        ("Check alerts for Tokyo", "weather alerts --city tokyo"),
    ],
    idempotent=True,
    see_also=["get", "forecast"],
)
def alerts(
    city: str = typer.Option("", help="Filter alerts by city (optional). type:string"),
    output: OutputFormat = typer.Option(
        OutputFormat.text, help="Output format. type:enum[text|json|table]"
    ),
) -> None:
    """List active weather alerts.

    Shows weather warnings and advisories. Optionally filter by city.
    """
    start = time.time()
    filtered = ALERTS
    if city:
        city = city.lower()
        if city not in CITIES:
            raise NotFoundError(f"Unknown city: '{city}'")
        filtered = [a for a in ALERTS if a["city"] == city]

    data = {"alerts": filtered, "count": len(filtered)}
    emit(success_envelope("alerts", data, version="1.0.0", start_time=start), output)


@app.command()
@acli_command(
    examples=[
        ("Add London to favorites", "weather favorite --city london"),
        ("Dry-run adding Paris", "weather favorite --city paris --dry-run"),
    ],
    idempotent="conditional",
    see_also=["get"],
)
def favorite(
    city: str = typer.Option(..., help="City to add to favorites. type:string"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview without saving. type:bool"),
    output: OutputFormat = typer.Option(
        OutputFormat.text, help="Output format. type:enum[text|json|table]"
    ),
) -> None:
    """Add a city to your favorites list.

    Saves a city for quick access. Use --dry-run to preview without saving.
    Idempotent when the city is already in favorites (no error, no duplicate).
    """
    start = time.time()
    city = city.lower()
    if city not in CITIES:
        raise NotFoundError(f"Unknown city: '{city}'")

    if dry_run:
        already = city in FAVORITES
        actions = [{"action": "add_favorite", "target": city, "reversible": True, "already_exists": already}]
        envelope = success_envelope(
            "favorite", {}, version="1.0.0", start_time=start, dry_run=True, planned_actions=actions
        )
        emit(envelope, output)
        raise SystemExit(ExitCode.DRY_RUN)

    if city not in FAVORITES:
        FAVORITES.append(city)

    data = {"city": city, "favorites": FAVORITES}
    emit(success_envelope("favorite", data, version="1.0.0", start_time=start), output)


@app.command()
@acli_command(
    examples=[
        ("Refresh all cities", "weather refresh"),
        ("Refresh specific cities", "weather refresh --cities london,paris"),
    ],
    idempotent=False,
    see_also=["get", "forecast"],
)
def refresh(
    cities: str = typer.Option(
        "", help="Comma-separated city names to refresh (default: all). type:string"
    ),
) -> None:
    """Refresh cached weather data for cities.

    Demonstrates NDJSON streaming for long-running operations.
    The --dry-run flag is auto-injected because idempotent=False.
    The --output flag is auto-injected by @acli_command.
    """
    target_cities = [c.strip() for c in cities.split(",") if c.strip()] if cities else list(CITIES)

    for c in target_cities:
        if c not in CITIES:
            raise NotFoundError(f"Unknown city: '{c}'")

    # Stream progress as NDJSON
    for c in target_cities:
        emit_progress("refresh", "running", detail=f"Fetching data for {c}")
        time.sleep(0.01)  # Simulated work

    emit_result({"cities_refreshed": target_cities, "count": len(target_cities)})


def main() -> None:
    """Entry point."""
    app.run()


if __name__ == "__main__":
    main()

File: examples/weather-ts/src/weather.ts
/**
 * weather — ACLI weather CLI (TypeScript port of examples/weather/weather.py).
 */

import {
  AcliApp,
  AcliError,
  ExitCode,
  type OutputFormat,
  dryRunEnvelope,
  emit,
  emitProgress,
  emitResult,
  successEnvelope,
} from '@acli/sdk';

const VERSION = '1.0.0';

const CITIES: Record<string, { lat: number; lon: number; country: string }> = {
  london: { lat: 51.5, lon: -0.1, country: 'GB' },
  paris: { lat: 48.9, lon: 2.3, country: 'FR' },
  tokyo: { lat: 35.7, lon: 139.7, country: 'JP' },
  'new-york': { lat: 40.7, lon: -74.0, country: 'US' },
  sydney: { lat: -33.9, lon: 151.2, country: 'AU' },
};

const ALERTS = [
  { city: 'tokyo', type: 'typhoon_warning', severity: 'high', message: 'Typhoon approaching' },
  { city: 'london', type: 'fog_advisory', severity: 'low', message: 'Dense fog expected' },
];

const FAVORITES: string[] = [];

function sortedCities(): string[] {
  return Object.keys(CITIES).sort();
}

function rngForCity(city: string): () => number {
  let s = [...city].reduce((acc, c) => acc * 31 + c.charCodeAt(0), 0) >>> 0;
  return () => {
    s = (Math.imul(s, 1103515245) + 12345) >>> 0;
    return (s & 0x7fffffff) / 0x7fffffff;
  };
}

function getWeather(city: string): Record<string, unknown> {
  const r = rngForCity(city);
  const conditions = ['sunny', 'cloudy', 'rainy', 'snowy', 'windy'];
  const info = CITIES[city];
  return {
    city,
    country: info.country,
    temperature_c: Math.round((r() * 40 - 5) * 10) / 10,
    humidity_pct: Math.floor(r() * 66) + 30,
    wind_kph: Math.round(r() * 50 * 10) / 10,
    condition: conditions[Math.floor(r() * 5)],
    coordinates: { lat: info.lat, lon: info.lon },
  };
}

function addImperial(data: Record<string, unknown>): Record<string, unknown> {
  const tc = data.temperature_c as number;
  const wk = data.wind_kph as number;
  return {
    ...data,
    temperature_f: Math.round(((tc * 9) / 5 + 32) * 10) / 10,
    wind_mph: Math.round(wk * 0.621 * 10) / 10,
  };
}

function forecast(city: string, days: number): Record<string, unknown> {
  const r = rngForCity(city);
  const info = CITIES[city];
  const conditions = ['sunny', 'cloudy', 'rainy', 'snowy'];
  const daily = [];
  for (let d = 0; d < days; d++) {
    daily.push({
      day: d + 1,
      high_c: Math.round((r() * 30 + 5) * 10) / 10,
      low_c: Math.round((r() * 25 - 5) * 10) / 10,
      condition: conditions[Math.floor(r() * 4)],
      precipitation_pct: Math.floor(r() * 101),
    });
  }
  return { city, country: info.country, days: daily };
}

function alertsJson(filter: string | undefined): Record<string, unknown> {
  const filtered =
    filter === undefined || filter === ''
      ? ALERTS
      : ALERTS.filter(a => a.city === filter);
  return {
    alerts: filtered,
    count: filtered.length,
  };
}

function buildApp(): AcliApp {
  const app = new AcliApp('weather', VERSION, { cliDir: process.cwd() });

  app
    .command(
      'get',
      `Get current weather for a city.

Returns temperature, humidity, wind speed, and conditions for the
specified city. Use --units imperial for Fahrenheit/mph.`,
      {
        examples: [
          ['Get weather for London', 'weather get --city london'],
          ['Get weather for Tokyo in JSON', 'weather get --city tokyo --output json'],
        ],
        idempotent: true,
        seeAlso: ['forecast', 'alerts'],
      },
      opts => {
        const output = opts.output as OutputFormat;
        const start = Date.now();
        const city = String(opts.city ?? '').toLowerCase();
        if (!CITIES[city]) {
          throw new AcliError(`Unknown city: '${city}'`, {
            code: ExitCode.NotFound,
            hint: `Available cities: ${sortedCities().join(', ')}`,
            docs: '.cli/examples/get.sh',
            command: 'get',
          });
        }
        let data: Record<string, unknown> = getWeather(city);
        if (String(opts.units ?? 'metric') === 'imperial') {
          data = addImperial(data);
        }
        emit(successEnvelope('get', data, VERSION, start), output);
      },
    )
    .option('--city <city>', 'City name (lowercase, hyphenated). type:string')
    .option('--units <units>', 'Unit system. type:enum[metric|imperial]', 'metric');

  app
    .command(
      'forecast',
      `Get multi-day weather forecast for a city.

Returns daily high/low temperatures and conditions for the next N days.`,
      {
        examples: [
          ['Get 3-day forecast for Paris', 'weather forecast --city paris --days 3'],
          [
            'Get 7-day forecast in JSON',
            'weather forecast --city london --days 7 --output json',
          ],
        ],
        idempotent: true,
        seeAlso: ['get', 'alerts'],
      },
      opts => {
        const output = opts.output as OutputFormat;
        const start = Date.now();
        const city = String(opts.city ?? '').toLowerCase();
        const days = Number(opts.days ?? 3);
        if (!CITIES[city]) {
          throw new AcliError(`Unknown city: '${city}'`, {
            code: ExitCode.NotFound,
            hint: `Available cities: ${sortedCities().join(', ')}`,
            command: 'forecast',
          });
        }
        if (days < 1 || days > 7) {
          throw new AcliError(`Days must be between 1 and 7, got ${days}`, {
            code: ExitCode.InvalidArgs,
            hint: 'Use --days with a value from 1 to 7',
            command: 'forecast',
          });
        }
        emit(successEnvelope('forecast', forecast(city, days), VERSION, start), output);
      },
    )
    .option('--city <city>', 'City name (lowercase, hyphenated). type:string')
    .option('--days <n>', 'Number of forecast days (1-7). type:int', '3');

  app
    .command(
      'alerts',
      `List active weather alerts.

Shows weather warnings and advisories. Optionally filter by city.`,
      {
        examples: [
          ['Check all active alerts', 'weather alerts'],
          ['Check alerts for Tokyo', 'weather alerts --city tokyo'],
        ],
        idempotent: true,
        seeAlso: ['get', 'forecast'],
      },
      opts => {
        const output = opts.output as OutputFormat;
        const start = Date.now();
        let filter: string | undefined;
        const raw = String(opts.city ?? '');
        if (raw !== '') {
          const c = raw.toLowerCase();
          if (!CITIES[c]) {
            throw new AcliError(`Unknown city: '${c}'`, {
              code: ExitCode.NotFound,
              command: 'alerts',
            });
          }
          filter = c;
        }
        emit(successEnvelope('alerts', alertsJson(filter), VERSION, start), output);
      },
    )
    .option('--city <city>', 'Filter alerts by city (optional). type:string', '');

  app
    .command(
      'favorite',
      `Add a city to your favorites list.

Saves a city for quick access. Use --dry-run to preview without saving.`,
      {
        examples: [
          ['Add London to favorites', 'weather favorite --city london'],
          ['Dry-run adding Paris', 'weather favorite --city paris --dry-run'],
        ],
        idempotent: 'conditional',
        seeAlso: ['get'],
      },
      opts => {
        const output = opts.output as OutputFormat;
        const start = Date.now();
        const city = String(opts.city ?? '').toLowerCase();
        if (!CITIES[city]) {
          throw new AcliError(`Unknown city: '${city}'`, {
            code: ExitCode.NotFound,
            command: 'favorite',
          });
        }
        if (opts.dryRun === true) {
          const planned = [
            {
              action: 'add_favorite',
              target: city,
              reversible: true,
              already_exists: FAVORITES.includes(city),
            },
          ];
          emit(dryRunEnvelope('favorite', planned, VERSION, start), output);
          process.exitCode = ExitCode.DryRun;
          return;
        }
        if (!FAVORITES.includes(city)) {
          FAVORITES.push(city);
        }
        emit(
          successEnvelope('favorite', { city, favorites: [...FAVORITES] }, VERSION, start),
          output,
        );
      },
    )
    .option('--city <city>', 'City to add to favorites. type:string')
    .option('--dry-run', 'Preview without saving. type:bool', false);

  app
    .command(
      'refresh',
      `Refresh cached weather data for cities.

Demonstrates NDJSON streaming for long-running operations.`,
      {
        examples: [
          ['Refresh all cities', 'weather refresh'],
          ['Refresh specific cities', 'weather refresh --cities london,paris'],
        ],
        idempotent: false,
        seeAlso: ['get', 'forecast'],
      },
      async opts => {
        const raw = String(opts.cities ?? '');
        const targetCities =
          raw === ''
            ? sortedCities()
            : raw
                .split(',')
                .map(s => s.trim().toLowerCase())
                .filter(Boolean);
        for (const c of targetCities) {
          if (!CITIES[c]) {
            throw new AcliError(`Unknown city: '${c}'`, {
              code: ExitCode.NotFound,
              command: 'refresh',
            });
          }
        }
        for (const c of targetCities) {
          emitProgress('refresh', 'running', `Fetching data for ${c}`);
          await new Promise(r => setTimeout(r, 10));
        }
        emitResult({ cities_refreshed: targetCities, count: targetCities.length }, true);
      },
    )
    .option(
      '--cities <cities>',
      'Comma-separated city names to refresh (default: all). type:string',
      '',
    );

  return app;
}

async function main(): Promise<void> {
  const app = buildApp();
  await app.run();
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});
